src/daemon/daemon.cpp

```diff
@@ -3315,37 +3344,62 @@ mp::MountHandler::UPtr mp::Daemon::make_mount(VirtualMachine* vm, const std::str
                : vm->make_native_mount_handler(config->ssh_key_provider.get(), target, mount);
 }
 
-void mp::Daemon::configure_new_interfaces(const std::string& name, mp::VirtualMachine& vm, mp::VMSpecs& specs)
+// This function configures the network interfaces whose MAC address is empty. They will stay in specs.extra_interfaces
+// only if the backend is able to add them to the virtual machine. If not, they will be removed.
+std::unordered_set<std::string> mp::Daemon::configure_new_interfaces(const std::string& name,
+                                                                     mp::VirtualMachine& vm,
+                                                                     mp::VMSpecs& specs)
 {
-    std::vector<size_t> new_interfaces;
+    bool added_good_interfaces{false};
+    size_t j = 0; /* The final index of the interface. This is needed because we will filter out the bad interfaces
+                     from specs.extra_interfaces. */
+    std::vector<mp::NetworkInterface> filtered_interfaces;
+    std::unordered_set<std::string> bad_bridges;
+    auto& commands = specs.run_at_boot;
 
-    for (auto i = 0u; i < specs.extra_interfaces.size(); ++i)
-    {
-        auto& interface = specs.extra_interfaces[i];
+    config->factory->prepare_networking(specs.extra_interfaces); // TODO: call this only if needed.
 
-        if (interface.mac_address.empty()) // An empty MAC address means the interface needs to be configured.
+    for (auto& iface : specs.extra_interfaces)
+    {
+        if (iface.mac_address.empty()) // An empty MAC address means the interface needs to be configured.
         {
-            new_interfaces.push_back(i);
+            iface.mac_address = generate_unused_mac_address(allocated_mac_addrs);
 
-            interface.mac_address = generate_unused_mac_address(allocated_mac_addrs);
-        }
-    }
+            if (vm.add_network_interface(j, iface))
+            {
+                added_good_interfaces = true;
 
-    if (!new_interfaces.empty())
-    {
-        config->factory->prepare_networking(specs.extra_interfaces);
+                commands.push_back(generate_netplan_script(j, iface.mac_address));
 
-        auto& commands = specs.run_at_boot;
+                filtered_interfaces.push_back(iface);
 
-        for (const auto i : new_interfaces)
+                ++j;
+            }
+            else
+            {
+                // The generated MAC will not be used, so we remove it from the list of used addresses.
+                // TODO: avoid adding the new MAC to allocated_mac_addrs if the interface is bad.
+                allocated_mac_addrs.erase(iface.mac_address);
+
+                bad_bridges.insert(iface.id);
```

Comment:
You could add a `continue` in the failure block and tighten up the code here. Something like this:
```c++
    for (auto& iface : specs.extra_interfaces)
    {
        if (iface.mac_address.empty()) // An empty MAC address means the interface needs to be configured.
        {
            iface.mac_address = generate_unused_mac_address(allocated_mac_addrs);

            if (vm.add_network_interface(j, iface))
            {
                added_good_interfaces = true;

                commands.push_back(generate_netplan_script(j, iface.mac_address));
            }
            else
            {
                // The generated MAC will not be used, so we remove it from the list of used addresses.
                // TODO: avoid adding the new MAC to allocated_mac_addrs if the interface is bad.
                allocated_mac_addrs.erase(iface.mac_address);

                bad_bridges.insert(iface.id);

                continue;
            }
        }

        filtered_interfaces.push_back(iface);

        j++;
    }
```

And if it's possible to use the `try`/`catch` I have proposed, it would look something like:
```c++
    for (auto& iface : specs.extra_interfaces)
    {
        if (iface.mac_address.empty()) // An empty MAC address means the interface needs to be configured.
        {
            iface.mac_address = generate_unused_mac_address(allocated_mac_addrs);

            try
            {
                vm.add_network_interface(j, iface);
   
                added_good_interfaces = true;

                commands.push_back(generate_netplan_script(j, iface.mac_address));
            }
            catch (const LXDNetworkError&)
            {
                // The generated MAC will not be used, so we remove it from the list of used addresses.
                // TODO: avoid adding the new MAC to allocated_mac_addrs if the interface is bad.
                allocated_mac_addrs.erase(iface.mac_address);

                bad_bridges.insert(iface.id);

                continue;
            }
        }

        filtered_interfaces.push_back(iface);

        j++;
    }
```

Reply:
That looks much better. Changed. Thanks!